infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msgHandlers/ReplicationHandlerMethods.java

```diff
@@ -0,0 +1,192 @@
+     */
+    protected HandlerMethod getRequestHandler(CorfuMessage.RequestMsg request) {
+        return requestHandlerMap.get(request.getPayload().getPayloadCase());
+    }
+
+    /**
+     * Given a {@link CorfuMessage.ResponseMsg} return the corresponding handler.
+     *
+     * @param response   The respose message to handle.
+     * @return          The appropriate {@link HandlerMethod}.
+     */
+    protected HandlerMethod getResponseHandler(CorfuMessage.ResponseMsg response) {
+        return responseHandlerMap.get(response.getPayload().getPayloadCase());
+    }
+
+    /**
+     * Handle an incoming Corfu request message.
+     *
+     * @param req       The request message to handle.
+     * @param r         The server router.
+     */
+    @SuppressWarnings("unchecked")
+    public void handle(CorfuMessage.RequestMsg req, CorfuMessage.ResponseMsg res, IClientServerRouter r) {
+        final HandlerMethod handler = req != null ? getRequestHandler(req) : getResponseHandler(res);
+
+        try {
+            handler.handle(req, res, r);
+        } catch (Exception e) {
+            if (req != null) {
+                log.error("handle[{}]: Unhandled exception processing {} request",
+                        req.getHeader().getRequestId(), req.getPayload().getPayloadCase(), e);
+            } else {
+                log.error("handle[{}]: Unhandled exception processing {} response",
+                        res.getHeader().getRequestId(), res.getPayload().getPayloadCase(), e);
+            }
+        }
+    }
+
+    /**
+     * Generate handlers for a particular server.
+     *
+     * @param caller    The context that is being used. Call MethodHandles.lookup() to obtain.
+     * @param server    The object that implements the server.
+     * @return          New request handlers for caller class.
+     */
+    public static ReplicationHandlerMethods generateHandler(@Nonnull final MethodHandles.Lookup caller,
+                                                            @NonNull final LogReplicationAbstractServer server) {
+        ReplicationHandlerMethods handler = new ReplicationHandlerMethods();
+        Arrays.stream(server.getClass().getDeclaredMethods())
+                .filter(method -> method.isAnnotationPresent(LogReplicationMsgHandler.class))
+                .forEach(method -> handler.registerMethod(caller, server, method));
+        return handler;
+    }
+
+    private void registerMethod(@Nonnull final MethodHandles.Lookup caller,
+                                @Nonnull final LogReplicationAbstractServer server,
+                                @Nonnull final Method method) {
+        final LogReplicationMsgHandler annotation = method.getAnnotation(LogReplicationMsgHandler.class);
+
+        // check if the method is already registered
+        if((!annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE) && requestHandlerMap.containsKey(annotation.requestType())) ||
```

Comment:
why we need this default payload case (NONE)?

Reply:
I see, annotation interface fields are mandatory...

actually I feel it won't be complicated to have two annotations right (request/response)? in the end we maintain 2 handler maps here, and we did the if check for request & response type...  since I feel it's a bit confusing to add that NONE type

Reply:
moving to having 2 annotations and removing NONE